api/rest/client/methods.go

```diff
@@ -172,3 +178,72 @@ func (c *Client) GetConnectGraph() (api.ConnectGraphSerial, error) {
 	err := c.do("GET", "/health/graph", nil, &graphS)
 	return graphS, err
 }
+
+// WaitFor is a utility function that allows for a caller to
+// wait for a paticular status for a CID. It returns a channel
+// upon which the caller can wait for the targetStatus.
+func (c *Client) WaitFor(ctx context.Context, ci *cid.Cid, local bool, target api.TrackerStatus, checkFreq time.Duration) (<-chan api.TrackerStatus, <-chan error) {
+	statusCh := make(chan api.TrackerStatus)
+	errCh := make(chan error)
+
+	go func() {
+		ticker := time.NewTicker(checkFreq)
+		defer ticker.Stop()
+
+	OUTER:
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				gblPinInfo, err := c.Status(ci, local)
+				if err != nil {
+					errCh <- err
+					return
+				}
+				for _, pinInfo := range gblPinInfo.PeerMap {
+					switch pinInfo.Status {
+					case target:
+						continue
+					case api.TrackerStatusBug, api.TrackerStatusClusterError, api.TrackerStatusPinError, api.TrackerStatusUnpinError:
+						errCh <- fmt.Errorf("error has occurred with cid: %s", ci.String())
+						statusCh <- pinInfo.Status
+						return
+					case api.TrackerStatusRemote:
+						if target == api.TrackerStatusPinned {
+							continue // to next pinInfo
+						}
+						statusCh <- pinInfo.Status
+						continue OUTER
+					default:
+						statusCh <- pinInfo.Status
+						continue OUTER
+					}
+				}
+				// NOTE: this code shouldn't be reached unless all peers have the target status
+				statusCh <- target
+				return
+			}
+		}
+	}()
+
+	return statusCh, errCh
+}
+
+// WaitForPinnedStatus ...
+func (c *Client) WaitForPinnedStatus(ctx context.Context, ci *cid.Cid, local bool, checkFreq time.Duration) (bool, error) {
```

Comment:
Taking into account all of the above, success here would mean:

* no errors received on error channel and no more items to read on it (because it's been closed)
* no more items to read from gPinInfo channel (because it's been closed too)

We might want to not return on errors, and just put them on the error channel and let the user decide whether to abort (via context) or keep waiting despite errors (maybe they're been addressed on the side). Let's leave that for next review round though.